polygeist/lib/Dialect/Polygeist/Transforms/FunctionSpecialization.cpp

```diff
@@ -0,0 +1,231 @@
+//===- FunctionSpecialization.cpp - Specialize functions ------------------===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+//
+// This pass attempts to version the callsite from SYCL kernel to the SYCL
+// kernel body function, where the specialized version ensures SYCL accessors do
+// not overlap.
+//
+//===----------------------------------------------------------------------===//
+
+#include "mlir/Dialect/Polygeist/Transforms/Passes.h"
+
+#include "mlir/Dialect/Arith/IR/Arith.h"
+#include "mlir/Dialect/Func/IR/FuncOps.h"
+#include "mlir/Dialect/GPU/IR/GPUDialect.h"
+#include "mlir/Dialect/Polygeist/IR/Ops.h"
+#include "mlir/Dialect/Polygeist/IR/Polygeist.h"
+#include "mlir/Dialect/Polygeist/Utils/TransformUtils.h"
+#include "mlir/Dialect/SYCL/IR/SYCLOps.h"
+#include "llvm/Support/Debug.h"
+
+#define DEBUG_TYPE "function-specialization"
+
+namespace mlir {
+namespace polygeist {
+#define GEN_PASS_DEF_FUNCTIONSPECIALIZATION
+#include "mlir/Dialect/Polygeist/Transforms/Passes.h.inc"
+} // namespace polygeist
+} // namespace mlir
+
+using namespace mlir;
+using AccessorType = VersionConditionBuilder::AccessorType;
+using AccessorPairType = VersionConditionBuilder::AccessorPairType;
+
+static llvm::cl::opt<unsigned> functionSpecializationAccessorLimit(
+    "function-specialization-accessor-limit", llvm::cl::init(5),
+    llvm::cl::desc(
+        "Maximum number of accessors allowed for function specialization"));
+
+//===----------------------------------------------------------------------===//
+// Helper Functions
+//===----------------------------------------------------------------------===//
+
+/// Returns true if \p type is 'memref<?x!sycl.accessor>', and false otherwise.
+static bool isValidMemRefType(Type type) {
+  auto mt = dyn_cast<MemRefType>(type);
+  bool isMemRefWithExpectedShape =
+      (mt && mt.hasRank() && (mt.getRank() == 1) &&
+       ShapedType::isDynamic(mt.getShape()[0]) && mt.getLayout().isIdentity());
+  if (!isMemRefWithExpectedShape)
+    return false;
+
+  return isa<sycl::AccessorType>(mt.getElementType());
+}
+
+/// Returns true if \p arg is a candidate argument.
+static bool isCandidateArgs(Value arg) {
+  if (!isa<AccessorType>(arg))
```

Comment:
Find it confusing because the name `AccessorType` implies we do not have a memref. Change name of the using declarations ?